services/spar/schema/src/V7.hs

```diff
@@ -0,0 +1,18 @@
+module V7 (migration) where
+
+import Imports
+import Cassandra.Schema
+import Text.RawString.QQ
+
+migration :: Migration
+migration = Migration 7 "Store default SSO code" $ do
+
+    -- partition_key_always_default should always be "default".
+    -- It exists so the row is always at a known partition.
```

Comment:
> Isn't the motivation here so that we guarantee to only have a single value?

No, that's not the case. I also tried that approach (before commit `schema composite primary key`/`bdc9396`), but what's currently here has `PRIMARY KEY (partition_key_always_default, idp)`, so you *can* theoretically have duplicates.

I had three slightly different schema versions I looked at and I could go back to the one with just `PRIMARY KEY (primary_key_always_default), but that makes the query for deleting an IdP more complicated (with an additional SELECT before you can DELETE) and I got errors from Cassandra (`java.lang.IndexOutOfBoundsException: readerIndex(207) + length(4) exceeds writerIndex(208)`) that I couldn't quite figure out.

> cassandra nodes themselves are aware which data resides where

As far as I understood from the reading I did last week, they only know that if they know the partitioning key, which here is `partition_key_always_default` (while `idp` is the clustering key). With the first schema I tried, without that "default" partition key, a query would look like `SELECT idp FROM default_idp`, so it doesn't specify the partition key, so the data could potentially be on any node.

Please correct me if my understanding is flawed.

So we basically have the three options:

1) single column `idp`
  - con: row could be in any partition, which is bad when querying it (the common case), although that apparently is not too bad with just 3 nodes?

2) additional column as primary key
  - pro: guarantees single row
  - con: requires additional SELECT with conditional DELETE on removing an IdP, comes with a race condition
  - also gave me errors (but I don't think it's a fundamental problem)

3) additional column as partitioning key, composite primary key
  - what we have now

Not sure if I forgot anything, but yeah, honestly not sure if going back to 2) is a good tradeoff. :man_shrugging: 

If my thoughts on partitioning keys don't make sense, 1) could also be a viable option again.